src/ray/core_worker/reference_count_test.cc

```diff
@@ -16,6 +19,124 @@ class ReferenceCountTest : public ::testing::Test {
   virtual void TearDown() {}
 };
 
+class MockWorkerClient : public rpc::CoreWorkerClientInterface {
+ public:
+  MockWorkerClient(ReferenceCounter &rc, const std::string &addr)
+      : rc_(rc), task_id_(TaskID::ForFakeTask()) {
+    address_.set_ip_address(addr);
+    address_.set_raylet_id(ClientID::FromRandom().Binary());
+    address_.set_worker_id(WorkerID::FromRandom().Binary());
+  }
+
+  // TODO: Make message receive and reply async.
+  ray::Status WaitForRefRemoved(
+      const rpc::WaitForRefRemovedRequest &request,
+      const rpc::ClientCallback<rpc::WaitForRefRemovedReply> &callback) override {
+    auto r = num_requests_;
+    requests_[r] = {
+        std::make_shared<rpc::WaitForRefRemovedReply>(),
+        callback,
+    };
+
+    auto send_reply_callback = [this, r](Status status, std::function<void()> success,
+                                         std::function<void()> failure) {
+      requests_[r].second(status, *requests_[r].first);
+    };
+    auto borrower_callback = [=]() {
+      rc_.SetRefRemovedCallback(request, requests_[r].first.get(), send_reply_callback);
+    };
+    borrower_callbacks_[r] = borrower_callback;
+
+    num_requests_++;
+    return Status::OK();
+  }
+
+  bool FlushBorrowerCallbacks() {
+    if (borrower_callbacks_.empty()) {
+      return false;
+    } else {
+      for (auto &callback : borrower_callbacks_) {
+        callback.second();
+      }
+      borrower_callbacks_.clear();
+      return true;
+    }
+  }
+
```

Comment:
Maybe leave a header comment that the below mirror the actual core worker operations and should be kept in sync